subprojects/core/src/integTest/groovy/org/gradle/api/ExternalScriptExecutionIntegrationSpec.groovy

```diff
@@ -73,4 +77,156 @@ task check {
+        errorOutput.contains("Could not get resource '${scriptUri}'")
+
+        when:
+        scriptFile.setText("""throw new RuntimeException('NOT CACHED')""", "UTF-8")
+        args('--offline', '-I', 'init.gradle')
+
+        then:
+        succeeds 'check'
+
+        where:
+        source        | sourceFile
+        "buildscript" | "build.gradle"
+        "settings"    | "settings.gradle"
+        "initscript"  | "init.gradle"
+    }
+
+    def "will only request resource once for build invocation"() {
+        given:
+        String scriptName = "script-${System.currentTimeMillis()}.gradle"
+        executer.withDefaultCharacterEncoding("ISO-8859-15")
+
+        and:
+        def scriptFile = file("script.gradle")
+        scriptFile.setText("""println 'loaded external script'""", "UTF-8")
+        server.expectGet('/' + scriptName, scriptFile)
+
+        and:
+        settingsFile << """
+            include ':projA'
+            include ':projB'
+        """
+
+        and:
+        [file('settings.gradle'), file('init.gradle'), file("projA/build.gradle"), file("projB/build.gradle")].each {
+            it << "apply from: 'http://localhost:${server.port}/${scriptName}'"
+        }
+
+        when:
+        args('-I', 'init.gradle')
+        succeeds 'tasks'
+
+        then:
+        output.count('loaded external script') == 4
+    }
+
+    def "will refresh cached value on subsequent build invocation"() {
+        given:
+        String scriptName = "script-${System.currentTimeMillis()}.gradle"
+        executer.withDefaultCharacterEncoding("ISO-8859-15")
+
+        and:
+        File scriptFile = file("script.gradle")
+        scriptFile.setText("""println 'loaded external script 1'""", "UTF-8")
+        scriptFile.makeOlder()
+
+        and:
+        buildFile << "apply from: 'http://localhost:${server.port}/${scriptName}'"
+
+        when:
+        server.expectGet('/' + scriptName, scriptFile)
+
+        then:
+        succeeds 'tasks'
+        output.contains('loaded external script 1')
+
+        when:
+        scriptFile.setText("""println 'loaded external script 2'""", "UTF-8")
+        server.expectHead('/' + scriptName, scriptFile)
+        server.expectGet('/' + scriptName, scriptFile)
+
+        then:
+        succeeds 'tasks'
+        output.contains('loaded external script 2')
+
+        when:
+        server.stop()
+        args("--offline")
+
+        then:
+        succeeds 'tasks'
+        output.contains('loaded external script 2')
+    }
+
+    def "reports and recovers from missing remote script"() {
+        String scriptName = "script-${System.currentTimeMillis()}.gradle"
+        String scriptUrl = "http://localhost:${server.port}/${scriptName}"
+        def scriptFile = file("script.gradle") << """
+            println 'loaded remote script'
+        """
+
+        buildFile << """
+            apply from: '${scriptUrl}'
+        """
+
+        when: // TODO:DAZ Should not be requesting this resource twice
```

Comment:
Because of https://github.com/gradle/gradle/blob/dd/url-cache/subprojects/dependency-management/src/main/java/org/gradle/internal/resource/transfer/DefaultUriTextResourceLoader.java#L48. I plan on addressing this.